### PR TITLE
Added a new retry mechenism to handle both in-session and out-session retries

### PR DESCRIPTION
**Pull request by yoroto**
*Issue #, if available:*

*Description of changes:*

Moved the retry logic into a RetryHandler class. Now it handles both in-session and out-session (previously driver level retries) retries in single place. All the retries are sharing the same retry limit.

### CODE REVIEW

File: Amazon.QLDB.Driver.Tests/QldbSessionTests.cs [hunk 1]
```diff
@@ -172,175 +171,103 @@ public void TestExecuteAbortExceptionAbortsTransactionAndRethrowsException()
...
 
             Assert.ThrowsException<InvalidSessionException>(() => qldbSession.Execute(
-                (TransactionExecutor txn) => { txn.Execute("testStatement"); return true; },
-                (int retry) => retryCount = retry));
-            Assert.AreEqual(3, txnCount);
-            Assert.AreEqual(0, commitCount);
-            Assert.AreEqual(2, retryCount);
+                (TransactionExecutor txn) => { txn.Execute("testStatement"); return true; }));
+
+            mockSession.Verify(s => s.End(), Times.Once);
```
Comment (byronlin-bq) [inline Amazon.QLDB.Driver.Tests/QldbSessionTests.cs:202]: Can we test the following logic
```
 this.isDisposed = true;
 this.disposeDelegate(null);
 this.Destroy();
 throw ise;
```

File: Amazon.QLDB.Driver.Tests/QldbSessionTests.cs [hunk 2]
```diff
@@ -172,175 +171,103 @@ public void TestExecuteAbortExceptionAbortsTransactionAndRethrowsException()
...
+                new object[] { new AmazonQLDBSessionException("", 0, "", "", HttpStatusCode.InternalServerError),
+                    typeof(RetriableException) },
+                new object[] { new AmazonQLDBSessionException("", 0, "", "", HttpStatusCode.ServiceUnavailable),
+                    typeof(RetriableException) },
+                new object[] { new AmazonQLDBSessionException("", 0, "", "", HttpStatusCode.Unauthorized),
+                    typeof(AmazonQLDBSessionException) },
+                new object[] { new OccConflictException("occ"),
+                    typeof(OccConflictException)},
+                new object[] { new AmazonServiceException(),
+                    typeof(AmazonServiceException)}
```
Comment (byronlin-bq) [inline Amazon.QLDB.Driver.Tests/QldbSessionTests.cs:269]: Can we add InvalidSessionException and TransactionAlreadyOpenException

File: Amazon.QLDB.Driver.Tests/RetryHandlerTests.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            Assert.IsFalse(handler.IsRetriable(new AmazonQLDBSessionException("aqse")));
+            Assert.IsFalse(handler.IsRetriable(new QldbDriverException("qldb")));
+
+            Assert.IsFalse(handler.NeedsRecover(new OccConflictException("occ")));
+            Assert.IsFalse(handler.NeedsRecover(new RetriableException(new Exception())));
+            Assert.IsTrue(handler.NeedsRecover(new InvalidSessionException("invalid")));
+        }
+
+        [TestMethod]
+        public void RetirableExecute_NoRetry_SuccessfulReturn()
```
Comment (byronlin-bq) [inline Amazon.QLDB.Driver.Tests/RetryHandlerTests.cs:31]: Retryable

File: Amazon.QLDB.Driver.Tests/RetryHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            Assert.IsTrue(handler.IsRetriable(new RetriableException(new Exception())));
+            Assert.IsTrue(handler.IsRetriable(new InvalidSessionException("invalid")));
+
+            Assert.IsFalse(handler.IsRetriable(new AmazonQLDBSessionException("aqse")));
+            Assert.IsFalse(handler.IsRetriable(new QldbDriverException("qldb")));
+
+            Assert.IsFalse(handler.NeedsRecover(new OccConflictException("occ")));
+            Assert.IsFalse(handler.NeedsRecover(new RetriableException(new Exception())));
+            Assert.IsTrue(handler.NeedsRecover(new InvalidSessionException("invalid")));
+        }
```
Comment (byronlin-bq) [inline Amazon.QLDB.Driver.Tests/RetryHandlerTests.cs:28]: Add TransactionAlreadyOpenException scenario

File: Amazon.QLDB.Driver/session/QldbSession.cs [hunk 5]
```diff
@@ -215,6 +176,7 @@ public virtual ITransaction StartTransaction()
             }
             catch (BadRequestException e)
             {
+                this.NoThrowAbort(null);
```
[the remark this answers is not in the document]
Comment (byronlin-bq) [reply Amazon.QLDB.Driver/session/QldbSession.cs:179]: What do you think of just updating:


```
        private void NoThrowAbort(ITransaction transaction)
        {
            try
            {
                    this.session.AbortTransaction();
            }
            catch (AmazonServiceException ase)
            {
                this.logger.LogWarning("Ignored error aborting transaction during execution: {}", ase);
```

File: Amazon.QLDB.Driver/session/QldbSession.cs [hunk 6]
```diff
@@ -124,78 +118,45 @@ public QldbSession Renew()
...
-                    this.NoThrowAbort(transaction);
-                    if (executionAttempt >= this.retryLimit ||
-                        ((aqse.StatusCode != HttpStatusCode.InternalServerError) &&
-                        (aqse.StatusCode != HttpStatusCode.ServiceUnavailable)))
-                    {
-                        throw aqse;
-                    }
+                transaction.Commit();
+                return returnedValue;
+            }
```
Comment (byronlin-bq) [inline Amazon.QLDB.Driver/session/QldbSession.cs:141]: I see in Node, it is catching `TransactionAbortedException`. As the name implies, we should catch this exception and call `this.NoThrowAbort()`

File: Amazon.QLDB.Driver/session/QldbSession.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (yoroto) [reply Amazon.QLDB.Driver/session/QldbSession.cs:179]: This is covered in Execute_ThrowBadRequestExceptionOnStartTransaction_ThrowTransactionOpenedException

File: Amazon.QLDB.Driver/session/QldbSession.cs (hunk 6 above)
Comment (yoroto) [reply Amazon.QLDB.Driver/session/QldbSession.cs:141]: this.NoThrowAbort() is called inside StartTransaction() where TransactionAbortedException was thrown. I changed it this way to simplify the logic.

Comment (byronlin-bq) [reply Amazon.QLDB.Driver/session/QldbSession.cs:141]: TransactionAbortedException is only thrown when user calls txn.abort() within a lambda, see TransactionExecutor.cs

File: Amazon.QLDB.Driver/session/QldbSession.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (byronlin-bq) [reply Amazon.QLDB.Driver/session/QldbSession.cs:179]: Awesome!

File: Amazon.QLDB.Driver/session/QldbSession.cs (hunk 6 above)
Comment (yoroto) [reply Amazon.QLDB.Driver/session/QldbSession.cs:141]: Yes, you are right. I will fixed it in the codecommit pr

**Comment by yoroto** [comment]
Not needed anymore

**Pull request closed without merging**